File: beta-hypergraphs.R
################################################################################
# Given a file of co-verticeships of the following format:
# vertex1,vertex2,vertex3
# vertex4
# vertex5,vertex6
# the function returns a list of vertices, and a list of hyperedges for each entry in filename.
# grep can be used from the terminal to easily preprocess a bib file to bring the article
# data to the required format.
################################################################################
Read.Hypergraph.From.File<-function(filename, simple = TRUE){
	H.file  = file(filename);    
	on.exit(close(H.file))
	hyperedges.str  = readLines(H.file);
	hyperedges = list()
	vertices = c();
	for (e in hyperedges.str){
		v.list = strsplit(e,split=',');
		# sort the hyperedges so that we can recognize multiple edges
		v.ordered.list = list(sort(unlist(v.list)))
		hyperedges = append(hyperedges, v.ordered.list)
		for (vertex in v.list){
			vertices = c(vertices, vertex)
		}
	}
	vertices = unique(vertices);
	if (simple){
		hyperedges = unique (hyperedges)	
	}
	return (list(vertices,hyperedges))
}

Get.Induced.Subgraph<-function(vertices, hyperedges, k){
	inducedH = list()
	for(e in hyperedges){
		if (all(e %in% vertices)){
				inducedH=append(inducedH,list(e))
		}
	}
	return(inducedH)
}

Get.Size.k.Hyperedges<-function(hyperedges, k){
# Given a list of hyperedges and an integer k returns the subset of the hyperedges
# that have size k.
	k.edges=list()
	if (is.list(k)){
		for (e in hyperedges){
			if (length(e) %in% k){
				k.edges = append(k.edges,list(e))
			}
		}	
	}
	else {
		for (e in hyperedges){
			if (length(e)==k){
				k.edges = append(k.edges,list(e))
			}
		}		
	}
	return(k.edges)
}

Get.Degree.Sequence<-function(hyperedges, vertices = NULL){
	if (is.null(vertices)){
		vertices = unique(unlist(hyperedges))	
	}
	n=length(vertices)
	d = rep(0,n)
	for (i in 1:n){ 
		for (e in hyperedges){
			if(vertices[i] %in% e){d[i]=d[i]+1}		
		}
	}
	return (d)
}

Get.Dense.Subgraph<-function(hyperedges, d.threshold=1, num.low.degree.vertices =0){
	d = Get.Degree.Sequence(hyperedges)
	vertices = unique(unlist(hyperedges))
	# Recursively remove all low degree vertices 
	# and edges that contain them from graph until the graph 
	# contains no low degree vertices or is empty
	while ( (length(vertices)>0) && (length(vertices[d<=d.threshold])>num.low.degree.vertices) ){
		vertices = vertices[d>1]
		hyperedges = Get.Induced.Subgraph(vertices, hyperedges)
		d = Get.Degree.Sequence(hyperedges)
	}
#	print("under construction")
	return(hyperedges)
}

Get.Smallest.Subgraph.Containing.Vertices<-function(vertices, hyperedges, k=NULL){
	subgraph = list()
	if (!is.null(k)){
		hyperedges = Get.Size.k.Hyperedges(hyperedges,k)
	}
	for(e in hyperedges){
		if (any(e %in% vertices)){
				subgraph=append(subgraph,list(e))
		}
	}
	return(subgraph)
}

Get.Smallest.Induced.Subgraph.Containing.Vertices<-function(vertices, hyperedges, k=NULL){
	new.vertex.set = list()
	if (!is.null(k)){
		hyperedges = Get.Size.k.Hyperedges(hyperedges,k)
	}
	for(e in hyperedges){
		if (any(e %in% vertices)){
				new.vertex.set=append(new.vertex.set,e)
		}
	}
	new.vertex.set = unique(new.vertex.set)
	return(Get.Induced.Subgraph(new.vertex.set, hyperedges,k=NULL))
}

# Use a fixed point algorithm to get the MLE
Estimate.Beta.Fixed.Point<-function(degrees, case="k-uniform", k, max.iter=500, tol=0.0001, beta = NULL){
	n=length(degrees)
	if (is.null(beta)){
		beta=rep(0,n)#log(degrees)#rep(0,n) #not sure what a good starting point would be
	}
	convergence=FALSE
	diff.list = list()
	steps=0
	#### There are more efficient methods for calculating e^{beta_S} for each S in n\choose k-1, e.g using a dynamic algorithm
	#sets = combinations(n,k-1,repeats.allowed=FALSE) #old code, defunct package?
	sets = t(combn(n,k-1))
	prod.exp.beta=rep(1,nrow(sets))
	while(convergence==FALSE && steps < max.iter){
		exp.beta=exp(beta) #calculate e^beta
		old.beta = beta
		if(any(is.infinite(old.beta))){
#			stop("Infinite beta estimate.")			
			return(NULL)
		}
		for (i in 1:nrow(sets) ){
			tuple=sets[i,] # tuple is a (k-1)-tuple of []1,2,...n]
#			print(tuple)
#			print(exp.beta[tuple])
			prod.exp.beta[i] = prod(exp.beta[tuple]) #calculate the product of e^beta for 
			if (is.infinite(prod.exp.beta[i])){
				print("Infinite beta.")
				return(NULL)
				}
		}
		# Calculate the beta estimate for each beta[i]
		##### currently very slow #####
		for (i in 1:n){
			sum.q.beta=0
			for (j in 1:nrow(sets)){
				tuple=sets[j,] # tuple is a (k-1)-tuple of [1,2,...n]	
				if (!(i %in% tuple)){
					sum.q.beta = sum.q.beta + prod.exp.beta[j]/(1+(prod.exp.beta[j]*exp.beta[i]))
					if (is.infinite(sum.q.beta)){
						print("Infinite beta.")
						return(NULL)
						}
				}	
				beta[i]=log(degrees[i])-log(sum.q.beta)
			}
		}
		diff = max(abs(old.beta-beta))
		print(paste("diff=", diff,  "-------- steps=",steps))
		print(beta)
		if (diff < tol){
			convergence=TRUE
		}
		steps=steps+1;
	}
	print(steps)
	print(diff)
	if (diff %in% diff.list){
		print("Divergent Subsequence Detected.")
		return(NULL)
		}
	diff.list = append(diff.list, list(diff))
	
	if (steps == max.iter){
		return(NULL)
		}
	else
		return (beta)
}
Estimate.Beta.Fixed.Point.General.Case<-function(degrees, k.list, max.iter=500, tol=1e-04, beta = NULL){
# Use a fixed point algorithm to get the MLE
# Accepts a list of edge-sizes 	
	n = length(degrees)
	if (is.null(beta)){ beta=rep(0,n)#log(degrees)#rep(0,n) #pick a starting point 
		}
	convergence=FALSE
	steps=0
	#### There are more efficient methods for calculating e^{beta_S} for each S in n\choose k-1, e.g using a dynamic algorithm
	#sets = combinations(n,k-1,repeats.allowed=FALSE) #old code, defunct package?
	all.index.sets = list()
	index.k = 0
	prod.exp.beta.list = list()
	for (k in k.list){
		print(k)
		index.k = index.k+1
		sets = t(combn(n,k-1))
		all.index.sets = append(all.index.sets, list(sets))
		prod.exp.beta.list=append(prod.exp.beta.list, list(rep(1, nrow(sets))))
	}
	
	diff.list = list()
	
	while(convergence==FALSE && steps < max.iter){
		exp.beta=exp(beta)
		old.beta = beta # To estimate convergence 
		if (any(is.infinite(old.beta))){
#			stop("Infinite beta estimate.")	
			print("Infinite beta.")
			return(NULL)
		}
		# Calculate e^{b_s} for all possible (k-1)-tuples
		for (index.k in 1:length(k.list)) {
			sets = all.index.sets[[index.k]]
			prod.exp.beta = prod.exp.beta.list[[index.k]]
			for (t in 1:nrow(sets) ){
				tuple=sets[t,] # tuple is a (k-1)-tuple of [1,2,...n] for some k in k.list
				prod.exp.beta[t] = prod(exp.beta[tuple])
				if (is.infinite(prod.exp.beta[t])){
					print("Infinite beta.")
					return(NULL)
				}
			}
			prod.exp.beta.list[[index.k]] = prod.exp.beta
#			print(paste("k = ", k.list[index.k], " complete.")) #testing
#			print(prod(prod.exp.beta)) #testing
		}
		# Calculate the beta estimate for each beta[i]
		for (i in 1:n) {
			sum.q.beta= rep(0, length(k.list))
			for (index.k in 1:length(k.list)){
				sets = all.index.sets[[index.k]]
				prod.exp.beta = prod.exp.beta.list[[index.k]]
				for (j in 1:nrow(sets)) {
					tuple=sets[j,] # tuple is a (k-1)-tuple of [1,2,...n]	for each k in k.list
					# See alternative code fragment in Estimate.Degree.Sequence.From.Beta and decide which is faster 
					if (!(i %in% tuple)) {
						sum.q.beta[index.k] = sum.q.beta[index.k] + prod.exp.beta[j]/(1+(prod.exp.beta[j]*exp.beta[i]))
						if (is.infinite(sum.q.beta[index.k])){
							print("Infinite beta.")
							return(NULL)
						}
					}					
				}
			}
			beta[i] = log(degrees[i]) - log(sum(sum.q.beta))
		}
		diff = max( abs(old.beta-beta) )
		print(paste("diff=", diff,  "-------- steps=",steps))
		print(beta)
		if (diff < tol){ convergence=TRUE }
		if (diff %in% diff.list){
			print("Divergent Subsequence Detected.")
			return(NULL)
			}
		diff.list = append(diff.list, list(diff))
		steps = steps+1;
	}
	print(steps)
	print(diff)
	if (steps == max.iter){return(NULL)}
	else return (beta)
}

Estimate.Degree.Sequence.From.Beta<-function(beta, k.list){
	#For testing purposes, layered k-uniform model
	n=length(beta)
	d = rep(0, n)
	exp.beta=exp(beta)
	for (i in 1:n){
		for (k in k.list){
			if(i==1){ tuples = combn(2:n,k-1) }
			else if(i==n){ tuples = combn(n-1,k-1) }
			else { tuples = combn(c(1:(i-1),(i+1):n),k-1) }
#alternative
#			apply(combn(n,k-1), 2, function(x) ifelse(i%in%x, beta.s=NULL, beta.s=beta[x]))			
			for (j in 1:ncol(tuples)){
				tuple = tuples[,j]
				exp.beta.tuple = exp(sum(beta[tuple]))
				d[i] = d[i] + (exp.beta.tuple*exp.beta[i]) / ( 1 + ( exp.beta.tuple*exp.beta[i] ) )
			}
		}
	}
	return(d)
}

Construct.IPS.Table<-function(vertices, hyperedges, model="k-uniform", k=2){
   print("UNDER CONSTRUCTION")
	n=length(vertices);
	t=array(0, dim=rep(n,k))	#initialize r-way array
#	t=rep(1/n,n^k);			#initialize r-way array
#	dim(t)<-c(rep(n,k)); 
	for (i in 1:n){
		t[rep(i,k)]=0;
		}
	
#	while(i<=3 && j<=4 && k<=5){ 
#		if ( i!=j && j!=l && i!=k){
#			v[i,j,k]=1/n;
#			}
#		}
	return(table)
}

# Construct a graph 
Reduce.To.Graph<-function(hyperedges, vertices=NULL, simple = TRUE){
	if (is.null(vertices)){
		vertices = unique(unlist(hyperedges))
	}
	edges = list()
	for (e in hyperedges){
		if (length(e) == 2){
			edges = append(edges, list(e))
		}
		else{
			pairs = combn(e,2) #returns matrix
			for (i in 1:ncol(pairs)){
				edges = append(edges, list(pairs[,i]))	
			}			
		}
	}
	if (simple){
		edges = unique (edges)	
	}
	return (list(vertices, edges))
}

Get.Probability.Under.k.uniform.Model<-function(beta, d, k, psi.beta = NULL){
	n=length(beta)
	if (is.null(psi.beta)){
		psi.beta=0
		tuples = combn(1:n,k)	
		for (j in 1:ncol(tuples)){
			tuple = tuples[,j]	
			psi.beta = psi.beta +log(1+ exp(sum(beta[tuple])))
		}
	}
	prob = exp(sum(d*beta)-psi.beta)
	return(prob)
}

Get.Probability.Under.Layered.Model<-function(beta.list, d.list, k.list, psi.beta.list = NULL){
	prob=1
	if (is.null(psi.beta.list)){ psi.beta.list=rep(NULL,length(k.list))}
	for (i in 1:length(k.list)){
		prob=prob * Get.Probability.Under.k.uniform.Model(beta.list[[i]], d.list[[i]],k.list[[i]], psi.beta.list[[i]])
	}
	return(prob)
}

Get.Probability.Under.General.Model<-function(beta, d, k.list, psi.beta = NULL){
	n=length(beta)
	if (is.null(psi.beta)){
		psi.beta=0
		for (k in k.list){
			tuples = combn(1:n,k)	
			for (j in 1:ncol(tuples)){
				tuple = tuples[,j]	
				psi.beta = psi.beta +log(1+ exp(sum(beta[tuple])))
			}
		}
	}
	prob = exp(sum(d*beta)-psi.beta)
	return(prob)
}

Get.Probability.Of.Edge.Under.Model<-function(beta, edge){
	exp.edge =exp(sum(beta[edge]))
	prob.edge = exp.edge/(1+exp.edge)
	return(prob.edge)
}

Draw.Random.Hypergraph.From.Model<-function(beta, k){
	hyperedges = list()
	n=length(beta)
	tuples = combn(1:n,k)
	for (i in 1:ncol(tuples)){
		tuple = tuples[,i]
		exp.sum.betas=exp(sum(beta[tuple]))
		prob.edge=(exp.sum.betas)/(1+ exp.sum.betas)
		coin = sample(c(TRUE,FALSE), size=1,prob=c(prob.edge,1-prob.edge))
		if (coin){hyperedges = append(hyperedges,list(tuple))}
	}
	return(hyperedges)
}

File: beta_hypergraphs.py
#!/usr/bin/env python3

import itertools
import time

import numba as nb
import numpy as np

from numba.typed import List


def deg_seq(hyperedges, vertices=None):
    """Get the sequence of degrees."""
    if vertices is None:
        vertices = list(set(itertools.chain.from_iterable(hyperedges)))

    n = len(vertices)
    d = np.zeros(n)
    for i in range(n):
        for e in hyperedges:
            if vertices[i] in e:
                d[i] += 1

    return d


def beta_fixed_point_R(degrees, k, sets, max_iter=500, tol=0.0001, beta=None):
    """Use a fixed point algorithm to calculate the MLE."""
    n = len(degrees)
    if beta is None:
        beta = np.zeros(n)

    convergence = False
    steps = 0

    # There are more efficient methods for calculating e^{beta_S} for each S in
    # n\choose k-1, e.g using a dynamic algorithm
    # sets = list(itertools.combinations(range(n), k - 1))
    prod_beta = np.ones(len(sets))

    while(not convergence and steps < max_iter):
        exp_beta = np.exp(beta)
        old_beta = beta.copy()
        if np.any(np.isinf(old_beta)):
            return None

        for i in range(len(sets)):
            prod_beta[i] = np.prod(np.array([exp_beta[i] for i in sets[i]]))

            if np.isinf(prod_beta[i]):
                print("Infinite beta")
                return

        for i in range(n):
            sum_q = 0
            for j in range(len(sets)):
                tt = sets[j]
                if i not in tt:
                    sum_q += prod_beta[j] / (1
                                             + prod_beta[j]
                                             * exp_beta[i])
                    if np.isinf(sum_q):
                        print("Infinite beta")
                        return

            beta[i] = np.log(degrees[i]) - np.log(sum_q)

        diff = np.max(np.abs(old_beta - beta))
        # print(f"diff= {diff} -------- steps= {steps}")
        # print(beta)
        if diff < tol:
            convergence = True

        steps += 1

    # print(steps)
    # print(diff)

    if steps == max_iter:
        return

    return beta


def beta_fixed_point(degrees, k, sets, max_iter=500, tol=0.0001, beta=None):
    """Use a fixed point algorithm to calculate the MLE."""
    n = len(degrees)
    sn = len(sets)
    an = np.arange(n)
    asn = np.arange(sn)
    if beta is None:
        beta = np.zeros(n)

    convergence = False
    steps = 0

    # There are more efficient methods for calculating e^{beta_S} for each S in
    # n\choose k-1, e.g using a dynamic algorithm
    # sets = np.asarray(list(itertools.combinations(range(n), k - 1)))

    ind = np.array([[j for j in asn if i not in sets[j]] for i in an])
    sets = np.asarray(sets)

    while(not convergence and steps < max_iter):
        exp_beta = np.exp(beta)
        old_beta = beta.copy()
        if np.any(np.isinf(old_beta)):
            return None

        # for i, s in enumerate(sets):
        #     prod_beta[i] = np.prod(exp_beta[np.asarray(s)])
        prod_beta = np.prod(exp_beta[sets], axis=1)

        if np.any(np.isinf(prod_beta)):
            print("Infinite beta")
            return

        # for i in range(n):
        #     ind = np.array([j for j in range(len(sets)) if i not in sets[j]])
        #     sum_q = np.sum(prod_beta[ind]
        #                    / (1 + prod_beta[ind] * exp_beta[i]))
        #     if np.isinf(sum_q):
        #         print("Infinite beta")

        sum_q = np.sum(prod_beta[ind] / (1 + (prod_beta[ind].T * exp_beta).T),
                       axis=1)

        if np.any(np.isinf(sum_q)):
            print("Infinite beta")
            return

        beta = np.log(degrees) - np.log(sum_q)

        diff = np.max(np.abs(old_beta - beta))
        # print(f"diff= {diff} -------- steps= {steps}")
        # print(beta)
        if diff < tol:
            convergence = True

        steps += 1

    # print(steps)
    # print(diff)

    if steps == max_iter:
        return

    return beta


def fixed_point_general_R(degrees, k_list, all_index_sets, max_iter=500,
                          tol=0.0001, beta=None):
    """Use a fixed point algorithm to get the MLE."""
    n = len(degrees)
    if beta is None:
        beta = np.zeros(n)

    convergence = False
    steps = 0

    prod_exp_beta_list = List()
    for s in all_index_sets:
        prod_exp_beta_list.append(np.ones(len(s)))

    while (not convergence and steps < max_iter):
        exp_beta = np.exp(beta)
        old_beta = beta.copy()
        if np.any(np.isinf(old_beta)):
            print("Infinite beta")
            return

        for index_k in range(len(k_list)):
            sets = all_index_sets[index_k]
            prod_exp_beta = prod_exp_beta_list[index_k]
            for t, s in enumerate(sets):
                eba = np.array([exp_beta[tt] for tt in s])
                prod_exp_beta[t] = np.prod(eba)
                if np.isinf(prod_exp_beta[t]):
                    print("Infinite beta")
                    return

            prod_exp_beta_list[index_k] = prod_exp_beta

        for i in range(n):
            sum_q_beta = np.zeros(len(k_list))
            for index_k in range(len(k_list)):
                sets = all_index_sets[index_k]
                prod_exp_beta = prod_exp_beta_list[index_k]
                for j in range(len(sets)):
                    if i not in sets[j]:
                        sum_q_beta[index_k] += (
                            prod_exp_beta[j]
                            / (1 + prod_exp_beta[j] * exp_beta[i])
                        )
                        if np.isinf(sum_q_beta[index_k]):
                            print("Infinite beta")
                            return
            beta[i] = np.log(degrees[i]) - np.log(np.sum(sum_q_beta))

        diff = np.max(np.abs(old_beta - beta))
        # print(f"diff= {diff} -------- steps= {steps}")
        # print(beta)
        if diff < tol:
            convergence = True

        steps += 1
    # print(steps)
    # print(diff)
    if steps == max_iter:
        print("Max iterations reached")
        return

    return beta


def fixed_point_general(degrees, k_list, all_index_sets, max_iter=500,
                        tol=0.0001, beta=None):
    """Use a fixed point algorithm to get the MLE."""
    n = len(degrees)
    if beta is None:
        beta = np.zeros(n)

    convergence = False
    steps = 0

    prod_exp_beta_list = list()
    for s in all_index_sets:
        prod_exp_beta_list.append(np.ones(len(s)))

    ns = np.asarray(all_index_sets)

    k_ind = range(len(k_list))
    ind = np.array([[[j for j in range(len(all_index_sets[k]))
                      if i not in all_index_sets[k][j]]]
                    for k in range(len(k_list))
                    for i in range(n)])

    while (not convergence and steps < max_iter):
        exp_beta = np.exp(beta)
        old_beta = beta.copy()
        if np.any(np.isinf(old_beta)):
            print("Infinite beta")
            return

        prod_exp_beta_list = np.prod(exp_beta[ns], axis=2)
        # for i in range(n):
        #     k_ind = range(len(k_list))
        #     ind = list(k_ind)

        #     # creates an "index" array
        #     for index_k in k_ind:
        #         iset = all_index_sets[index_k]
        #         ind[index_k] = np.array([j for j in range(len(iset))
        #                                  if i not in iset[j]])

        #     sum_q_beta = np.sum(
        #         prod_exp_beta_list[k_ind, ind]
        #         / (1 + prod_exp_beta_list[k_ind, ind] * exp_beta[i])
        #     )
        #     beta[i] = np.log(degrees[i]) - np.log(sum_q_beta)

        sum_q_beta = np.sum(prod_exp_beta_list[k_ind, ind]
                            / (1 + prod_exp_beta_list[k_ind, ind].T
                               * exp_beta).T, axis=2)
        # TODO don't know why it's a list of lists
        sum_q_beta = sum_q_beta[:, 0]
        beta = np.log(degrees) - np.log(sum_q_beta)

        diff = max(abs(old_beta - beta))
        # print(f"diff= {diff} -------- steps= {steps}")
        # print(beta)
        if diff < tol:
            convergence = True

        steps += 1
    # print(steps)
    # print(diff)
    if steps == max_iter:
        print("Max iterations reached")
        return

    return beta


def main():
    # for correctness
    n = 5
    k = 3
    K53 = list(itertools.combinations(range(n), k))
    degs = deg_seq(K53)
    sets = List(itertools.combinations(range(len(degs)), k - 1))
    fp_njit = nb.njit(beta_fixed_point_R)
    fpg_njit = nb.njit(fixed_point_general_R)

    beta_K53 = beta_fixed_point(degs, k=k, sets=sets, max_iter=10000)
    all_passed = np.allclose(beta_K53, 3.07028833 * np.ones(5))
    print(f"Testing correctness for n={n}, k={k}; passing={all_passed}")
    print()

    print(f"Precompiling python jit'd code (with n={n}, k={k})")
    tic = time.perf_counter()
    fp_njit(degs, k=k, sets=sets, max_iter=10000)
    toc = time.perf_counter()
    print(f"beta_fixed_point jit'd took {toc - tic:0.4f} seconds")
    print()

    # for performance
    n = 25
    k = 3
    Knk = list(itertools.combinations(range(n), k))
    degs = deg_seq(Knk)
    sets = List(itertools.combinations(range(len(degs)), k - 1))

    print(f"Running R-converted code (with n={n}, k={k})")
    tic = time.perf_counter()
    beta_fixed_point_R(degs, k=k, sets=sets, max_iter=10000)
    toc = time.perf_counter()
    print(f"beta_fixed_point_R took {toc - tic:0.4f} seconds")
    print()

    # print(beta_Kn3)

    print(f"Running python vectorized code (with n={n}, k={k})")
    tic = time.perf_counter()
    beta_fixed_point(degs, k=k, sets=sets, max_iter=10000)
    toc = time.perf_counter()
    print(f"beta_fixed_point took {toc - tic:0.4f} seconds")
    print()

    print(f"Running python jit'd code (with n={n}, k={k})")
    tic = time.perf_counter()
    fp_njit(degs, k=k, sets=sets, max_iter=10000)
    toc = time.perf_counter()
    print(f"beta_fixed_point_R jit'd took {toc - tic:0.4f} seconds")
    # print()

    # d10_3 = (36, 36, 36, 36, 36)
    # n = len(d10_3)
    # k_list = List([k, ])
    # all_index_sets = List()
    # for k in k_list:
    #     sets = List(itertools.combinations(range(n), k - 1))
    #     all_index_sets.append(sets)

    # print(f"Precompiling python jit'd code (with n={n}, k={k_list})")
    # tic = time.perf_counter()
    # fpg_njit(d10_3, k_list, all_index_sets, max_iter=10000)
    # toc = time.perf_counter()
    # print(f"fixed_point_general jit'd took {toc - tic:0.4f} seconds")
    # print()

    # d10_3 = (36, 36, 36, 36, 36, 36, 36, 36, 36, 36)
    # n = len(d10_3)
    # k_list = List([k, ])
    # all_index_sets = List()
    # for k in k_list:
    #     sets = List(itertools.combinations(range(n), k - 1))
    #     all_index_sets.append(sets)

    # print(f"Running python R-converted code (with n={n}, k={k_list})")
    # tic = time.perf_counter()
    # fixed_point_general_R(d10_3, List(k_list), all_index_sets, max_iter=10000)
    # toc = time.perf_counter()
    # print(f"fixed_point_general_R took {toc - tic:0.4f} seconds")
    # print()

    # print(f"Running python vectorized code (with n={n}, k={k_list})")
    # tic = time.perf_counter()
    # fixed_point_general(d10_3, k_list, all_index_sets, max_iter=10000)
    # toc = time.perf_counter()
    # print(f"fixed_point_general took {toc - tic:0.4f} seconds")
    # print()

    # # [3.07028833 3.07028833 3.07028833 3.07028833 3.07028833 3.07028833
    # #  3.07028833 3.07028833 3.07028833 3.07028833]
    # # 3331
    # # 9.997253637461512e-05

    # print(f"Running python jit'd code (with n={n}, k={k_list})")
    # tic = time.perf_counter()
    # fpg_njit(d10_3, k_list, all_index_sets, max_iter=10000)
    # toc = time.perf_counter()
    # print(f"fixed_point_general_R jit'd took {toc - tic:0.4f} seconds")


if __name__ == "__main__":
    main()
